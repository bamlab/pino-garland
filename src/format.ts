import chalk from "chalk";
import { center, padNumber } from "./utils/text";

export abstract class Format {
  abstract canFormat(logData: LogData): boolean;
  abstract format(logData: LogData): string;
  abstract placeholder(): string;
}

export class FormatDate extends Format {
  private SIZE: number;

  constructor() {
    super();
    const placeholderTimestamp = new Date(0).getTime();
    this.SIZE = this.format({ time: placeholderTimestamp }).length;
  }

  canFormat(logData: { time?: number }): boolean {
    return logData.time !== undefined;
  }

  format(logData: { time: number }): string {
    const { time } = logData;
    const formattedTime = new Date(time);

    const hours = padNumber(formattedTime.getHours(), 2);
    const minutes = padNumber(formattedTime.getMinutes(), 2);
    const seconds = padNumber(formattedTime.getSeconds(), 2);
    const milliseconds = padNumber(formattedTime.getMilliseconds(), 3);

    return chalk.gray(`${hours}:${minutes}:${seconds}.${milliseconds}`);
  }

  placeholder(): string {
    return " ".repeat(this.SIZE);
  }
}

export class FormatRequestId extends Format {
  private REQUEST_ID_SIZE = 18;
  private SIZE: number = this.REQUEST_ID_SIZE + 5;

  constructor() {
    super();
  }

  canFormat(logData: { req?: { id?: string } }): boolean {
    return logData.req !== undefined && logData.req.id !== undefined;
  }

  format(logData: { req: { id: string } }): string {
    const requestId = logData.req.id;
    const shortRequestId = requestId.substring(0, this.REQUEST_ID_SIZE);
    return chalk.cyan(`[req: ${shortRequestId}]`);
  }

  placeholder(): string {
    const placeholderRequestId = center("no request", this.SIZE);
    return chalk.grey(`[${placeholderRequestId}]`);
  }
}
