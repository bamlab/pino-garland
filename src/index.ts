export { parse } from "./parse";
export { FormatDate, FormatRequestId } from "./format";
export { print, printUnknown } from "./print";
