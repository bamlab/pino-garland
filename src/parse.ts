import { ParseError } from "./errors/ParseError";
import { getErrorMessage } from "./utils/getErrorMessage";

export function parse(logJsonLine: string): LogDataWithLevelName {
  try {
    const logData = JSON.parse(logJsonLine);

    if (logData.level === undefined) {
      throw new ParseError("Log line doesn't have a level");
    }

    if (typeof logData.level === "number") {
      return replaceLevelWithLevelName(logData);
    }

    return logData;
  } catch (err) {
    const message = getErrorMessage(err);
    throw new ParseError(message);
  }
}

function replaceLevelWithLevelName(logData: LogDataWithLevelNumber): LogDataWithLevelName {
  let level: string;
  if (logData.level === 10) level = "trace";
  else if (logData.level === 20) level = "debug";
  else if (logData.level === 30) level = "info";
  else if (logData.level === 40) level = "warn";
  else if (logData.level === 50) level = "error";
  else if (logData.level === 60) level = "fatal";
  else level = "unknown";
  return { ...logData, level };
}
