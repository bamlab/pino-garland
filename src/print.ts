import { EOL } from "node:os";
import { Format, FormatDate, FormatRequestId } from "./format";

const formatters: Format[] = [new FormatDate(), new FormatRequestId()];

export function print(logData: LogData): string {
  const parts = [];

  for (const formatter of formatters) {
    if (formatter.canFormat(logData)) {
      parts.push(formatter.format(logData));
    } else {
      parts.push(formatter.placeholder());
    }
  }

  return `${parts.join(" ")}${EOL}`;
}

export function printUnknown(jsonLine: string): string {
  return `${jsonLine}${EOL}`;
}
