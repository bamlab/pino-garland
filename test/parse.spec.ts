import { parse } from "../src/parse";

describe("parse", () => {
  it("should parse a log line", () => {
    // Given
    const logLine = '{"level": 30, "time": 1539619829389, "msg": "hello"}';
    // When
    const logData = parse(logLine);
    // Then
    expect(logData).toEqual({
      level: "info",
      time: 1539619829389,
      msg: "hello",
    });
  });

  it("should throw an error if the log line is not valid JSON", () => {
    // Given
    const logLine = "not valid JSON";
    // When
    const parseLogLine = () => parse(logLine);
    // Then
    expect(parseLogLine).toThrowErrorMatchingSnapshot();
  });

  it("should throw an error if the log line does not have a level", () => {
    // Given
    const logLine = '{"time": 1539619829389, "msg": "hello"}';
    // When
    const parseLogLine = () => parse(logLine);
    // Then
    expect(parseLogLine).toThrowErrorMatchingSnapshot();
  });

  it("should replace the log line's level with a level name", () => {
    // Given
    const logLine = '{"level": 30, "time": 1539619829389, "msg": "hello"}';
    // When
    const logData = parse(logLine);
    // Then
    expect(logData).toEqual({
      level: "info",
      time: 1539619829389,
      msg: "hello",
    });
  });
});
