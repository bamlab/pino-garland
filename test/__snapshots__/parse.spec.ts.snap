// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse should throw an error if the log line does not have a level 1`] = `"Log line doesn't have a level"`;

exports[`parse should throw an error if the log line is not valid JSON 1`] = `"Unexpected token o in JSON at position 1"`;
